src/Agent.Worker/Build/BuildJobExtension.cs

```diff
@@ -29,25 +29,33 @@ public BuildJobExtension()
                 @finally: false);
         }
 
-        private async Task<TaskResult> PrepareAsync()
+        private Task PrepareAsync()
         {
+            // Validate args.
             Trace.Entering();
-            var directoryManager = HostContext.GetService<IBuildDirectoryManager>();
+            ArgUtil.NotNull(PrepareStep, nameof(PrepareStep));
+            ArgUtil.NotNull(PrepareStep.ExecutionContext, nameof(PrepareStep.ExecutionContext));
             IExecutionContext executionContext = PrepareStep.ExecutionContext;
+            var directoryManager = HostContext.GetService<IBuildDirectoryManager>();
+
+            // Get the repo endpoint and source provider.
             ServiceEndpoint endpoint = null;
             ISourceProvider sourceProvider = null;
             if (!TryGetPrimaryEndpointAndSourceProvider(executionContext, out endpoint, out sourceProvider))
             {
-                executionContext.Error(StringUtil.Loc("SupportedRepositoryEndpointNotFound"));
-                return TaskResult.Failed;
+                throw new Exception(StringUtil.Loc("SupportedRepositoryEndpointNotFound"));
```

Comment:
I figured out an error handling strategy across the worker process. Now task-handlers and job-plugins can throw. This should make it easy for the task-handler and job-plugin implementations to communicate failure and control flow.